### ISSUE TITLE
ADT Enum Wrong Typeclass Inference

### ISSUE THREAD

**Issue by AugustNagro** [lampepfl/dotty#11381]
## Compiler version

3.0.0-M3

## Minimized code

Lets say I am making a type class JsonCodec:

```Scala
import scala.deriving._
import scala.compiletime.{erasedValue, summonInline, constValue}

trait JsonCodec[T]:
  extension (t: T) def toJson: String

object JsonCodec:
  inline def apply[T](using jc: JsonCodec[T]) = jc

  given JsonCodec[String] with
    extension (t: String) def toJson: String = t
  
  inline given derived[T](using m: Mirror.Of[T]): JsonCodec[T] = inline m match
    case _: Mirror.ProductOf[T] =>
      new JsonCodec[T]:
        println("created new JsonCodec for PRODUCT")
        extension (t: T) def toJson: String =
          constValue[m.MirroredLabel].toString
    case _: Mirror.SumOf[T] =>
      new JsonCodec[T]:
        println("created new JsonCodec for SUM")
        extension (t: T) def toJson: String =
          constValue[m.MirroredLabel].toString
```

Then, I make a enum Opt, and try to serialize it:

```Scala
enum Opt[+T] derives JsonCodec:
  case Sm(theValue: T)
  case N

val o: Opt.Sm[String] = Opt.Sm("hello world")

println(o.toJson) // prints 'Opt'

println(JsonCodec[Opt.Sm[String]].toJson(o)) // prints 'Sm'
```

## Expectation

I would expect that `println(o.toJson)` uses the extension method of the most-specialized given JsonCodec.. ie the one in Opt.Sm's companion object.

Although maybe my understanding is wrong.

**Comment by bishabosha** [comment]
@AugustNagro So when you put a `derives` clause on an enum, it puts into the companion object `Opt` a single instance `derived$JsonCodec[T]` which is computed by calling `JsonCodec.derived[Opt[T]]`, which has a `SumOf` typed mirror. When you call an arbitrary extension method on `o: Opt.Sm[String]`, It will start looking in lexical and implicit scope for eligible extension methods called `toJson`, which finds that computed derived instance in `Opt` companion object. At this point `JsonCodec` companion is not even searched.

The second example is more "precise" because you are looking for an implicit `JsonCodec[Opt.Sm[String]]` value, so the implicit scope will include `JsonCodec` companion object and prioritise the `JsonCodec.derived` given, which it calls by passing the Mirror for `Opt.Sm[String]` which is more precise.

So to summarise, the extension method is less precise because we are looking for any extension called `toJson`, whereas the second example is more precise because you are looking for the type first.

**Comment by AugustNagro** [comment]
@bishabosha Your explanation makes 100% sense.

I think the approach to derived serialization of enum ADTs is then

1. For sum case, use ordinal() to summon the JsonCodec for the product subclass
2. For the product case, return the type name if instance of Mirror.Singleton, otherwise serialize using given JsonCodecs for the product elements.
